jetty-websocket/jetty-websocket-client/src/main/java/org/eclipse/jetty/websocket/client/impl/JettyClientUpgradeRequest.java

```diff
@@ -43,16 +45,24 @@
 {
     private final WebSocketClient containerContext;
     private final Object websocketPojo;
-    private final CompletableFuture<Session> futureSession;
     private final DelegatedJettyClientUpgradeRequest handshakeRequest;
+    private final CompletableFuture<Session> completableFutureSession;
+    private final CompletableFuture<Session> futureSession;
+    private UpgradeResponse upgradeResponse;
 
     public JettyClientUpgradeRequest(WebSocketClient clientContainer, WebSocketCoreClient coreClient, UpgradeRequest request,
                                      URI requestURI, Object websocketPojo)
     {
         super(coreClient, requestURI);
         this.containerContext = clientContainer;
         this.websocketPojo = websocketPojo;
-        this.futureSession = new CompletableFuture<>();
+        this.completableFutureSession = new CompletableFuture<>();
+        this.futureSession = completableFutureSession.thenApply(s->
+        {
+            // Note: we must set the upgrade response before the future session will complete
+            ((WebSocketSession)s).setUpgradeResponse(upgradeResponse);
+            return s;
+        });
```

Comment:
Seems to me you don't need `completableFutureSession`, but only `futureSession`?